### ISSUE TITLE
lockFileURL fails at resolving from the underlying frozen object

### ISSUE THREAD

**Issue by WebReflection** [pyodide/pyodide#5135]
### To Reproduce

Try to use `jsonpointer==3.0.0` or anything else as pinned version as config for your `lockFileURL` logic and see that [this is not enough](https://github.com/pyodide/pyodide/blob/2aa31871e4c22182c3945df5baa92b12b3d0ca00/src/js/packaging-utils.ts#L14) to eventually map back that version to a pinned, frozen, locked file, where the version is embedded in its details, not in the name.

That `canonicalizePackageName` function doesn't take into account pinned versions at all so it fails out of the box when pinned versions where used in the config.

```js
export function canonicalizePackageName(name: string): string {
  return name.split('==')[0].replace(canonicalizeNameRegex, "-").toLowerCase();
}
```

Above would fix already the issue but I don't have enough background to understand why that function, involved when `lockFileURL` is passed as option, would just lower-case modules and forget people might have pinned versions within every package name.

### Expected behavior

passing a `micropip.frozen()` resulting *JSON* after a module pinned with a version should *not* fail out of the blue saying:

`Uncaught (in promise) Error: No known package with name 'jsonpointer==3.0.0'`

### Environment

- Pyodide Version: latest
- Browser version: latest
- Any other relevant information: [this issue elsewhere](https://github.com/pyscript/pyscript/issues/2228)

**Comment by Archmonger** [comment]
Notably, another issue I'm seeing with the current behavior is it is impossible to ever upgrade a dependency without forcing the user to clear indexdb.

**Comment by hoodmane** [comment]
@Archmonger that sounds like a downstream problem. Pyodide doesn't touch indexdb itself.

**Comment by hoodmane** [comment]
[I said on the other issue:](https://github.com/pyscript/pyscript/issues/2228#issuecomment-2438205493)

> Yeah we don't have any logic for parsing [requirement specifiers](https://packaging.pypa.io/en/stable/specifiers.html) in `loadPackage`. We only load locked requirements based on an exact name match. You should strip the specifier part of the requirement out yourself. I would do: 
> ```js
> function requirementToName(requirement) {
>    return requirement.split(/[^A-Za-z0-9_-]/,1)[0];
> }
> ```
> Perhaps we should explain the workflow for freezing requirements somewhere in our docs...

Another thing we might consider is providing this `requirementToName()` as a utility function so we can guide downstream towards correct application. The regex is actually pretty simple and I think it's right but it's nontrivial to validate so it might be helpful to provide it. I think it makes sense to ask the downstream to do `packages: Array.from(requirementList, requirementToName)`. WDYT @ryanking13?

**Comment by ryanking13** [comment]
> packages: Array.from(requirementList, requirementToName). WDYT @ryanking13?

Hmm, I think it is better to handle it internally in loadPackage than asking the caller to do it.

The reason why we were not dealing with this was mostly because we didn't want to re-implement the pacakging library in JS... but probably it is worth to consider some options.

I think we can either impelement the packaging specifiers in JS or vendor packaging library. The simple regex will not cover comparing versions, extras, etc.

**Comment by WebReflection** [comment]
Maybe my example wasn't self-explanatory ... but it's possible I finally understand where we should strip out versions: not as key, just as packages needed when frozen packages are already found ... am I correct about this assumption?

Here a more concise example:

```html
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width,initial-scale=1.0">
  <script type="module">
    const url = 'https://cdn.jsdelivr.net/pyodide/v0.26.3/full/pyodide.mjs';
    const indexURL = url.slice(0, url.lastIndexOf('/'));
    const options = { indexURL };
    const packages = ['jsonpointer==3.0.0'];
    const alreadyStored = localStorage.getItem('packages');
    if (alreadyStored) {
      console.log('Reusing frozen packages');
      const blob = new Blob([alreadyStored], { type: 'application/json' });
      options.lockFileURL = URL.createObjectURL(blob);
      // IS IT HERE WHERE I SHOULD NORMALIZE NAMES ???
      options.packages = packages;
    }
    const { loadPyodide } = await import(url);
    const interpreter = await loadPyodide(options);
    if (!options.lockFileURL) {
      console.log('Creating frozen packages');
      await interpreter.loadPackage('micropip');
      const micropip = interpreter.pyimport('micropip');
      await micropip.install(packages, { keep_going: true });
      localStorage.setItem('packages', micropip.freeze());
      micropip.destroy();
    }
    interpreter.runPython('print("Hello Pyodide")');
  </script>
</head>
</html>
```

If I run it first time everything is fine but if I refresh the page it fails the same way ... **however** if I use this instead:

```js
options.packages = packages.map(name => name.split('==')[0]);
```

I can indeed refresh and have no issues, all seem to be working correctly and one could change the `runPython` code to something like:

```js
interpreter.runPython('import jsonpointer; print("OK")');
```

and zero issues found ... if that's it, I think I am OK in landing a workaround but I still consider this a workaround because:

  * unless I've missed it, nowhere in the current docs is specified that frozen packages can't be retrieved with the same name that as used to freeze these earlier
  * *loadPyodide* receives an *optional* `locakFileURL` and I think it should be responsibility of that optional flag to normalize packages because users might feel confused about the fact they did in fact used a specific packages config to define modules and these suddenly break the bootstrap if reused
  * on the other hand, I don't think you should add too much logic but at least ignoring the version, as is not part of the JSON exported name, would be less surprise prone?

I think I am going, for the time being, to normalize what I can normalized because currently we suggest to pin versions to our users and then who does that has a fully broken PyScript experience so the sooner we fix this the better, but don't you think it would be kinda nice to expect things to "*just work*" and versioning being better handled in *pyodide* when a `lockFileURL` is provided, hoping that won't delay more its bootstrap time?

Thanks in anyway to both @hoodmane and @ryanking13 for providing your always valuable feedback.

**Comment by WebReflection** [comment]
FYI I've applied the workaround suggested by @hoodmane in the right place and we're out also with the latest *pyodide* as we missed somehow the release and we were lacking behind a patch.

Everything now seems to work just fine on our side so there's no urgency in fixing this on your side but I think something should be done, either at the documentation level with a proper example, and you can feel free to reuse my page example code to do so (although that might open a can of worms because `localStorage` is limited and not suitable for more complex locked in dependencies) or tell users that when `lockFileURL` is used the `packages` used with `loadPyodide` should **not** contain the pinned versions in it, otherwise it fails at bootstrapping (but maybe it fails regardless, here I must admit I've always used `micropip` instead to bootstrap packages, it looks like it was a fortunate choice after all).

**Comment by hoodmane** [comment]
Yeah adding a page in the documentation page showing a workflow to lock dependencies is a good idea. I think using `localStorage` makes perfect sense for a docs page, the point isn't to explain how to do persistent storage correctly, it's to illustrate how to work with Pyodide. Usually when I try to make these things, I find out that things are broken and don't even work the way I hoped...

**Comment by WebReflection** [comment]
> I think using localStorage makes perfect sense for a docs page

that's why I've used it in my example, it's the least thing to explain as storage for the Web so it worked fine there ... *but* there's gonna be a tiny *hint* that for bigger graphs, the 5MB granted by localStorage might not be enough and other ways to store the frozen *micropip* JSON might be more suitable.

A workflow around this topic though is what I was missing from the beginning, glad if I've helped sketching one down in here :-)

**Comment by ryanking13** [comment]
> @ryanking13 perhaps a good starting point would be to check for characters outside of [a-zA-Z0-9_-] and throw an error saying that loadPackage() doesn't process specifiers? Another option would indeed be to allow package_name==version_locked as an alternative, but it would be good to keep it simple to begin with.

Yeah, I think we can start with raising errors when special characters are passed to `loadPackage`, and think of a better way to handle this.

**Comment by eyupcanakman** [comment]
Still happens on current main: `loadPackage(["jsonpointer==3.0.0"])` throws `No known package with name 'jsonpointer==3.0.0'`. I'll put up a small PR for the first step you agreed on (reject specifiers with a clear error), leaving the broader parsing for later.
